### PR TITLE
general improvements to `tensor`

### PR DESCRIPTION
**Pull request by samuelpmishLLNL**
This PR is meant to break up some of the changes that were piling up in another branch. Some highlights of this PR are:
- changing some of serac's low level classes (`tensor`, `dual`, etc) to not rely on the nvcc flag `--expt-relaxed-constexpr` to function, since there are some compiler bugs associated with enabling that flag.
- adding a new class, `isotropic_tensor` that makes arithmetic between isotropic and dense tensors use less memory and perform fewer calculations.
- expanding `tensor_unit_tests` and `tensor_unit_tests_cuda` and updating them to use `gtest`
- modify the `tensor` class itself to use a more general, and more succinct implementation
- reimplementing a few functions that gcc-11 was compiling incorrectly (see https://github.com/LLNL/serac/issues/603#issuecomment-983012969)

### CODE REVIEW

File: src/serac/numerics/functional/tensor.hpp [hunk 1]
```diff
@@ -1576,36 +1337,46 @@ auto inv(tensor<dual<gradient_type>, n, n> A)
...
   return out;
 }
 
 /**
- * @brief recursively serialize the entries in a tensor to an ostream.
- * Output format uses braces and comma separators to mimic C syntax for multidimensional array
- * initialization.
- *
- * @param[in] out the std::ostream to write to (e.g. std::cout or std::ofstream)
+ * @brief print a doulbe using `printf`, so that it is suitable for use inside cuda kernels. (used in final recursion of
```
Comment (jandrej) [inline src/serac/numerics/functional/tensor.hpp]: s/doulbe/double

**Review by jandrej — APPROVED** [review]

File: src/serac/numerics/functional/isotropic_tensor.hpp [hunk 2]
```diff
@@ -0,0 +1,457 @@
...
+
+  T value;  ///< the value on the diagonal
+};
+
+/**
+ * @brief return the identity matrix of the specified size
+ * @tparam m the number of rows and columns
+ */
+template <int m>
+SERAC_HOST_DEVICE constexpr isotropic_tensor<double, m, m> Identity()
```
Comment (jamiebramwell) [inline src/serac/numerics/functional/isotropic_tensor.hpp:61]: This will be nice to have.

File: src/serac/numerics/functional/isotropic_tensor.hpp [hunk 3]
```diff
@@ -0,0 +1,457 @@
...
+
+/**
+ * @brief compute the squared Frobenius norm (tr(dot(transpose(I), I))) of an isotropic tensor
+ *
+ * @tparam T the types stored in the isotropic tensor
+ * @tparam m the number of rows and columns in I
+ * @param I the isotropic tensor to compute the squared norm of
+ */
+template <typename T, int m>
+SERAC_HOST_DEVICE constexpr auto sqnorm(const isotropic_tensor<T, m, m>& I)
```
Comment (jamiebramwell) [inline src/serac/numerics/functional/isotropic_tensor.hpp]: Is this common shorthand for this? If so, I'm fine with it. Otherwise, I think a more verbose name (`squared_norm`) is more appropriate.

File: src/serac/numerics/functional/isotropic_tensor.hpp [hunk 4]
```diff
@@ -0,0 +1,457 @@
...
+ * @tparam S the types stored in isotropic tensor
+ * @tparam T the types stored in tensor
+ * @tparam m the dimension of each extent of I
+ * @param I the left operand
+ * @param A the (full) right operand
+ * @return a new tensor equal to the index notation expression:
+ *    output(...) := I(i,j) * A(i,j,...)
+ */
+template <typename S, typename T, int m, int... n>
+SERAC_HOST_DEVICE constexpr auto ddot(const isotropic_tensor<S, m, m, m, m>& I, const tensor<T, m, m, n...>& A)
```
Comment (jamiebramwell) [inline src/serac/numerics/functional/isotropic_tensor.hpp]: Same comment as `sqnorm`. If this is common (e.g. a Matlab/Mathematica call), I'm fine with it. Otherwise, it should be more explicit (e.g. `double_dot`).

File: src/serac/physics/solid_functional.hpp [hunk 5]
```diff
@@ -575,7 +575,7 @@ class SolidFunctional : public BasePhysics {
   mfem::Vector zero_;
 
   /// @brief Auxilliary identity rank 2 tensor
-  const tensor<double, dim, dim> I_ = Identity<dim>();
+  const isotropic_tensor<double, dim, dim> I_ = Identity<dim>();
```
Comment (jamiebramwell) [inline src/serac/physics/solid_functional.hpp:578]: 🎉

File: src/serac/physics/thermal_conduction_functional.hpp [hunk 6]
```diff
@@ -153,10 +153,12 @@ class ThermalConductionFunctional : public BasePhysics {
 
     functional_call_args_.emplace_back(temperature_.trueVec());
 
-    for (size_t i = 0; i < sizeof...(parameter_space); ++i) {
-      trial_spaces[i + 1]         = &(parameter_states_[i].get().space());
-      parameter_sensitivities_[i] = std::make_unique<FiniteElementDual>(mesh_, parameter_states_[i].get().space());
-      functional_call_args_.emplace_back(parameter_states_[i].get().trueVec());
+    if constexpr (sizeof...(parameter_space)) {
```
Comment (jamiebramwell) [inline src/serac/physics/thermal_conduction_functional.hpp:156]: Why is this needed?

File: src/serac/numerics/functional/isotropic_tensor.hpp (hunk 3 above)
Comment (samuelpmishLLNL) [reply src/serac/numerics/functional/isotropic_tensor.hpp]: not that I know of, some of these things were just stolen from some old code of mine back in grad school. I think `squared_norm` is better (or do we want this function at all?)

File: src/serac/physics/thermal_conduction_functional.hpp (hunk 6 above)
Comment (samuelpmishLLNL) [reply src/serac/physics/thermal_conduction_functional.hpp:156]: oops, I was a getting compiler error inside the `for` loop (just below the line of code you highlighted):

```cpp
for (size_t i = 0; i < sizeof...(parameter_space); ++i) {
                                      ^
```

When `parameter_space` is empty, then the end condition on the for loop is essentially just the literal `0`.
The compiler was complaining the comparison `i < 0` made the loop trivial.

File: src/serac/numerics/functional/isotropic_tensor.hpp (hunk 3 above)
Comment (jamiebramwell) [reply src/serac/numerics/functional/isotropic_tensor.hpp]: I'm OK either deleting it or renaming it.

**Review by jamiebramwell — APPROVED** [review]

File: src/serac/infrastructure/profiling.cpp [hunk 7]
```diff
@@ -20,6 +20,12 @@ std::optional<cali::ConfigManager> mgr;
 }  // namespace
 #endif
 
+/**
+ * @brief initialization for the profiling libraries (currently adiak and caliper, when enabled)
```
Comment (white238) [inline src/serac/infrastructure/profiling.cpp:24]: Thanks!

File: src/serac/numerics/functional/isotropic_tensor.hpp (hunk 4 above)
Comment (samuelpmishLLNL) [reply src/serac/numerics/functional/isotropic_tensor.hpp]: hah, I was trying to remember where this came from, it's LaTeX for 2 dots over a character (not a useful name here).

File: src/serac/numerics/functional/isotropic_tensor.hpp [hunk 8]
```diff
@@ -0,0 +1,457 @@
...
+template <typename T, int... n>
+struct isotropic_tensor;
+
+/**
+ * @brief there is no such thing as a rank-1 isotropic tensor, but we include this specialization
+ * to help explain that to users, rather than just producing an "incomplete type" compilation error
+ */
+template <typename T, int n>
+struct isotropic_tensor<T, n> {
+  static_assert(::detail::always_false<T>{}, "error: there is no such thing as a rank-1 isotropic tensor!");
```
Comment (white238) [inline src/serac/numerics/functional/isotropic_tensor.hpp:40]: 👍

**Review by white238 — APPROVED** [review]
